cohttp/src/proxy.ml

```diff
@@ -0,0 +1,145 @@
+      and b = String.(sub b 0 (min (length b) n) |> lowercase_ascii) in
+      String.compare a b = 0
+
+    let no_proxy_from_env_value no_proxy =
+      match no_proxy with
+      | None -> Patterns []
+      | Some no_proxy ->
+          if no_proxy = "*" then Wildcard
+          else
+            let patterns =
+              no_proxy
+              |> String.split_on_char ','
+              |> List.filter_map (fun pattern ->
+                     if pattern = "" then None else Some (String.trim pattern))
+              |> List.map (fun pattern ->
+                     match Ipaddr.of_string pattern with
+                     | Ok addr -> Ipaddr_prefix (Ipaddr.Prefix.of_addr addr)
+                     | Error _ -> (
+                         match Ipaddr.Prefix.of_string pattern with
+                         | Ok prefix -> Ipaddr_prefix prefix
+                         | Error _ ->
+                             Name (trim_dots ~first_leading:true pattern)))
+            in
+            Patterns patterns
+
+    let check_no_proxy uri pattern =
+      let host = Uri.host_with_default ~default:"" uri in
+      match pattern with
+      | Wildcard -> true
+      | _ when String.length host = 0 -> true
+      | Patterns patterns -> (
+          match Ipaddr.of_string host with
+          | Ok hostip ->
+              List.exists
+                (function
+                  | Name _ -> false
+                  | Ipaddr_prefix network -> Ipaddr.Prefix.mem hostip network)
+                patterns
+          | Error _ ->
+              let name = trim_dots ~first_leading:false host in
+              List.exists
+                (function
+                  | Ipaddr_prefix _ -> false
+                  | Name pattern ->
+                      let patternlen = String.length pattern
+                      and namelen = String.length name in
+                      if patternlen = namelen then
+                        (* An exact (case-insensitive) match *)
+                        strncasecompare pattern name namelen
+                      else if patternlen < namelen then
+                        (* pattern is a (case-insensitive) suffix of the host,
+                         starting after any subdomain prefix.
+
+                         E.g., [example.com] is a suffix of [www.example.com] and
+                         [home.example.com], but not of [nonexample.com]. *)
+                        let match_start = namelen - patternlen in
+                        let host_suffix =
+                          String.sub name match_start patternlen
+                        in
+                        name.[match_start - 1] = '.'
+                        && strncasecompare pattern host_suffix patternlen
+                      else false)
+                patterns)
+
+    type ('direct, 'tunnel) t = Direct of 'direct | Tunnel of 'tunnel
+
+    type ('direct, 'tunnel) servers = {
+      by_scheme : (string * ('direct, 'tunnel) t) list;
```

Comment:
For improved readability maybe it would make sense to introduce `type scheme = string` here? It's a bit of a shame that the scheme is represented just as a string.

Reply:
Yes, that's a valid point. I quite dislike the fact that the inference can take the type and would suddenly infer other strings to this, just because of the alias.